### PR TITLE
add makefile targets to install codespell and verify using codespell

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
- Adds two makefile `targets`
  - `make codespell` installs `codespell` binary in the _hack/tools/bin/codespell/_ dir
  - `make verify-codespell` runs `$(CODESPELL) $(ROOT_DIR) --ignore-words=$(ROOT_DIR)/.codespellignore --skip="*.git,*_artifacts,*.sum,$(ROOT_DIR)/hack"` at the root of the codebase. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
- `make verify` should help users in correcting codespell errors before raising a PR. 

**Special notes for your reviewer**:
- You may notice that codespell gets installed in `hack/tools/bin/codespell` directory. That is because installing codespell brings along other dependencies. I tried using `pip --no-deps install .. codespell` but could not get rid of other dependencies you see in the picture below.
    Screenshot 2023-06-21 at 2 59 20 PM

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nawazkh** [comment]
Hmm.. I dont see this in my local. 
```
.
.
./hack/verify-starlark.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

100 4680k  100 4680k    0     0  7233k      0 --:--:-- --:--:-- --:--:-- 7233k
Running buildifier...
Cleaning up...
which /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/codespell >/dev/null || ( \
        mkdir -p /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin; \
        pip install --target=/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/codespell codespell; \
    )
Collecting codespell
  Downloading codespell-2.2.5-py3-none-any.whl (242 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 242.7/242.7 kB 5.7 MB/s eta 0:00:00
Installing collected packages: codespell
Successfully installed codespell-2.2.5
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
 [notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: python3 -m pip install --upgrade pip
Traceback (most recent call last):
  File "/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/codespell/bin/codespell", line 5, in <module>
    from codespell_lib import _script_main
ModuleNotFoundError: No module named 'codespell_lib'
make: *** [Makefile:273: verify-codespell] Error 1
+ EXIT_VALUE=2
+ set +o xtrace
Cleaning up after docker in docker.
```


My local output on `make verify`
```
.
.
./hack/verify-starlark.sh
Running buildifier...
Cleaning up...
which /Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell >/dev/null || ( \
        mkdir -p /Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin; \
        pip install --target=/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell codespell; \
    )
Collecting codespell
  Using cached codespell-2.2.5-py3-none-any.whl (242 kB)
Installing collected packages: codespell
Successfully installed codespell-2.2.5
```

File: Makefile [hunk 1]
```diff
@@ -841,3 +851,11 @@ $(SETUP_ENVTEST_BIN): $(SETUP_ENVTEST) ## Build a local copy of setup-envtest.
 
 $(SETUP_ENVTEST): # Build setup-envtest from tools folder.
 	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) $(SETUP_ENVTEST_PKG) $(SETUP_ENVTEST_BIN) $(SETUP_ENVTEST_VER)
+
+$(CODESPELL): ## Build codespell from tools folder.
+	@which $(CODESPELL) >/dev/null || ( \
+        mkdir -p $(TOOLS_BIN_DIR); \
+        pip install --target=$(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR) $(CODESPELL_BIN)==$(CODESPELL_VER); \
+		mv $(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR)/bin/$(CODESPELL_BIN) $(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR); \
+		rm -r $(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR)/bin; \
```
Comment (nawazkh) [inline Makefile:860]: I had to move `codespell` from `$(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR)/bin/$(CODESPELL_BIN)` to `$(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR)` to avoid running into `ModuleNotFoundError: No module named 'codespell_lib'` errors.

**Comment by nawazkh** [comment]
cc @mboersma

**Comment by mboersma** [comment]
I think this makes sense, because it's not a good experience to have your PR blocked by a linter that you can't run locally first.

It looks good to me overall. I'll try it out.

**Comment by jackfrancis** [comment]
If we ship this change we'll want to ship it simultaneously with a removal of the github action, otherwise we're running it twice on each PR (see the `pull-cluster-api-provider-azure-verify` job that ran as part of this PR, for example).

I assume the reason we've been doing the github action up until now is because no one wanted to do the chop wood / carry water work (until @nawazkh now, I mean!) of integrating this into our local CI interfaces?

**Comment by nawazkh** [comment]
> If we ship this change we'll want to ship it simultaneously with a removal of the github action, otherwise we're running it twice on each PR (see the pull-cluster-api-provider-azure-verify job that ran as part of this PR, for example).

I think that's a fair ask, thanks for pointing me to it. 
I am in favor of removing the GitHub action, because the PR's `make verify-codespell` achieves the same as done in the Github action. In addition to that, `pull-cluster-api-provider-azure-verify` is most likely to block a codespell error than a GitHub action.

**Comment by nawazkh** [comment]
Commit 8b5d3536282182f0976195c480f44acddbe041f9 removes the GitHub code flow.

File: Makefile [hunk 2]
```diff
@@ -841,3 +851,11 @@ $(SETUP_ENVTEST_BIN): $(SETUP_ENVTEST) ## Build a local copy of setup-envtest.
 
 $(SETUP_ENVTEST): # Build setup-envtest from tools folder.
 	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) $(SETUP_ENVTEST_PKG) $(SETUP_ENVTEST_BIN) $(SETUP_ENVTEST_VER)
+
+$(CODESPELL): ## Build codespell from tools folder.
+	@which $(CODESPELL) >/dev/null || ( \
+        mkdir -p $(TOOLS_BIN_DIR); \
+        pip install --target=$(TOOLS_BIN_DIR)/$(CODESPELL_DIST_DIR) $(CODESPELL_BIN)==$(CODESPELL_VER); \
```
Comment (CecileRobertMichon) [inline Makefile:858]: We should add pip to the developer doc as a pre-requirement https://capz.sigs.k8s.io/developers/development.html#base-requirements since some people may not have it pre-installed locally

**Review by CecileRobertMichon — APPROVED** [review]
Tried this locally and it worked great, thanks @nawazkh! One small comment about adding pip to developer docs

File: Makefile (hunk 2 above)
Comment (nawazkh) [reply Makefile:858]: Good point! Updating the documentation.

Comment (nawazkh) [reply Makefile:858]: Updated!

File: Makefile [hunk 3]
```diff
@@ -264,6 +269,10 @@ verify-conversions: $(CONVERSION_VERIFIER)  ## Verifies expected API conversion
 verify-tiltfile: ## Verify Tiltfile format.
 	./hack/verify-starlark.sh
 
+.PHONY: verify-codespell
+verify-codespell: codespell ## Verify codespell.
+	@$(CODESPELL) $(ROOT_DIR) --ignore-words=$(ROOT_DIR)/.codespellignore --skip="*.git,*_artifacts,*.sum,$(ROOT_DIR)/hack"
```
Comment (CecileRobertMichon) [inline Makefile:274]: sorry, one last question: why did you add the hack dir here? I see in the github action that it only had `.git,_artifacts,*.sum` as skip setting

Comment (nawazkh) [reply Makefile:274]: Good question, sorry, I should have specified in the PR description about it.
I added `$(ROOT_DIR)/hack` to avoid codespell errors upon checking `cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/tests/test_basic.py`
Removing `$(ROOT_DIR)/hack` from the `verify-codespell` target leads to a bunch of errors like below. 

```
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:54: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:55: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:56: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:57: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:58: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:59: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:60: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:61: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:62: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:63: ECT ==> ETC
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:68: whic ==> which
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:72: ect ==> etc
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:73: infinit ==> infinite
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:74: infinit ==> infinite
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:76: ISNT ==> ISN'T
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:77: Teh ==> The
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:78: Teh ==> The
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:79: Teh ==> The
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:80: Teh ==> The
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:81: pallete ==> palette
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:82: pallete ==> palette
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:83: pallete ==> palette
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:84: virtualY ==> virtually
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:85: archType ==> archetype
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:86: Satelite ==> Satellite
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/linux-kernel.exclude:87: SMOE ==> SOME
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/dictionary_usage.txt:9: useability ==> usability
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell_lib/data/dictionary_usage.txt:17: useability ==> usability
/Users/nawazhussain/msftcode/cluster-api-provider-azure/hack/tools/bin/codespell_dist/codespell-2.2.5.dist-info/METADATA:55: adn ==> and
```

Comment (nawazkh) [reply Makefile:274]: But now that I think about it, I can further fine-tune just to remove codespell_dist folder. Updating.

Comment (nawazkh) [reply Makefile]: Updated! :)

File: Makefile (hunk 2 above)
Comment (CecileRobertMichon) [reply Makefile:858]: thank you!

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @mboersma 

I really like that we're no longer using codespell @ head with this change so we won't be broken by codespell updates like we have in the past

**Comment by CecileRobertMichon** [comment]
@nawazkh any idea why the broken lints github action is failing?

**Comment by nawazkh** [comment]
> @nawazkh any idea why the broken lints github action is failing?

Looks like it is pointing to some broken links. I have seen such linter links fix themselves up, but I am not sure if I should update them this time.
```

FILE: ./docs/book/src/topics/ssh-access.md
[✖] https://azure.microsoft.com/en-us/services/azure-bastion/

3 links checked.

ERROR: 1 dead links found!
[✖] https://azure.microsoft.com/en-us/services/azure-bastion/ → Status: 0

FILE: ./docs/book/src/topics/custom-dns.md

1 links checked.

FILE: ./docs/book/src/topics/getting-started.md
[✖] https://azure.microsoft.com/en-us/

9 links checked.

ERROR: 1 dead links found!
[✖] https://azure.microsoft.com/en-us/ → Status: 0
```

**Comment by nawazkh** [comment]
Ready for review! :)

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

Tested locally, `make verify-codespell` seems to work as advertised. Thanks @nawazkh!
